docs/advanced/runtime/docs/atomics-memory-model.md

```diff
@@ -0,0 +1,131 @@
+guarantees to hold.
+
+If a load or store to an address happens at the same time as another load or store to that address but of a different
+size, all bets are off and no atomicity is guaranteed.
+
+These rules apply to high-level languages like C# and F# as they target the CLI.
+
+### `volatile.` prefix opcode in CIL
+
+When the `volatile.` prefix opcode is used in CIL, it imposes acquire/release semantics on the next non-prefix opcode.
+For loads, it results in acquire semantics. For stores, it results in release semantics.
+
+This prefix opcode has no effect on atomicity beyond the standard rules of the CLI.
+
+### `volatile` keyword in C\#
+
+The `volatile` keyword in C# compiles down to `Thread.VolatileRead` and `Thread.VolatileWrite` calls. Nothing more,
+nothing less. This means loads or stores to `volatile` fields have the same semantics as these methods.
+
+C#'s `volatile` cannot be applied to 64-bit quantities because these methods do not guarantee atomicity for 64-bit
+quantities on 32-bit systems, and the `Volatile` class did not exist when the `volatile` keyword was designed. Today,
+`volatile` on 64-bit quantities could conceivably be compiled down to `Volatile.Read` and `Volatile.Write` calls.
+
+### `Thread` class
+
+The `VolatileRead` and `VolatileWrite` methods perform loads and stores with acquire and release semantics,
+respectively. They guarantee absolutely nothing about atomicity beyond the standard rules of the CLI. In effect, this
+means that the 64-bit overloads of these methods are not atomic on 32-bit systems.
+
+There is a quirk in the Microsoft implementation where these methods actually use the `MemoryBarrier` method to insert
+a barrier. This is stronger than a simple acquire or release barrier. We do the same for compatibility.
+
+The `MemoryBarrier` method inserts a full sequential consistency barrier.
+
+### `Volatile` class
+
+The methods on the `Volatile` class are all atomic regardless of system bitness, and result in acquire and release
+barriers for loads and stores respectively.
+
+The 64-bit methods on this class are not atomic with respect to loads or stores made through other means than the
+methods on this class and the `Interlocked` class. This is because such 64-bit operations may need to be implemented
+with a lock on 32-bit systems.
+
+The MSDN documentation incorrectly states that the C# compiler emits calls to this class's methods. It does not. This
+statement is probably intentional as, disregarding the implementation quirk with the `Thread` methods described above,
+C#'s `volatile` is actually supposed to behave the way that the methods on this class do. So the intention is likely to
+avoid having new code rely on the quirk in the `Thread` methods.
+
+### `Interlocked` class
+
+The methods on this class all have much the same semantics as those on the `Volatile` class, with the exception that
```

Comment:
I think this description is confusing. It's much simpler to say "The methods on this class are always atomic WRT to each other and those in the Volatile class. Method on this class impose sequential consistency."

We should state all cases in terms of atomicity and memory ordering.